src/constraints/RebarBondSlipConstraint.C

```diff
@@ -0,0 +1,348 @@
+    if (_bondslip.find(it->first) == _bondslip.end())
+      _bondslip.insert(std::pair<dof_id_type, bondSlipData>(it->first,
+                                                            bondSlipData())); // initialize
+}
+
+void
+RebarBondSlipConstraint::timestepSetup()
+{
+  for (auto iter = _slave_to_master_map.begin(); iter != _slave_to_master_map.end(); ++iter)
+  {
+    dof_id_type node_id = iter->first;
+    auto it = _bondslip.find(node_id);
+    mooseAssert(it != _bondslip.end(), "Node not found in bond-slip map");
+
+    _bondslip[node_id].slip_min_old = _bondslip[node_id].slip_min;
+    _bondslip[node_id].slip_max_old = _bondslip[node_id].slip_max;
+    _bondslip[node_id].bondstress_min_old = _bondslip[node_id].bondstress_min;
+    _bondslip[node_id].bondstress_max_old = _bondslip[node_id].bondstress_max;
+  }
+}
+
+bool
+RebarBondSlipConstraint::shouldApply()
+{
+  if (_debug)
+  // if (_current_node->id() == 144)
+  {
+    std::cout << "===========================================\n";
+    std::cout << "node id: " << _current_node->id() << std::endl;
+    std::cout << "at coord: " << (Point)*_current_node << std::endl;
+  }
+  auto it = _slave_to_master_map.find(_current_node->id());
+
+  if (it != _slave_to_master_map.end())
+  {
+    const Elem * master_elem = _mesh.elemPtr(it->second);
+    std::vector<Point> points = {*_current_node};
+
+    // reinit variables on the master element at the slave point
+    _fe_problem.setNeighborSubdomainID(master_elem, 0);
+    _fe_problem.reinitNeighborPhys(master_elem, points, 0);
+
+    reinitConstraint();
+
+    return true;
+  }
+  return false;
+}
+
+void
+RebarBondSlipConstraint::computeTangent()
+{
+  _slave_tangent *= 0.0;
+
+  // get normals
+  // get connected elements of the current node
+  const std::map<dof_id_type, std::vector<dof_id_type>> & node_to_elem_map = _mesh.nodeToElemMap();
+  auto node_to_elem_pair = node_to_elem_map.find(_current_node->id());
+  mooseAssert(node_to_elem_pair != node_to_elem_map.end(), "Missing entry in node to elem map");
+  const std::vector<dof_id_type> & elems = node_to_elem_pair->second;
+
+  for (auto & elem : elems)
+  {
+    Elem * elem_ptr = _mesh.elemPtr(elem);
+    // _assembly.reinit(elem_ptr, 0);
+    _current_elem_volume += _assembly.elemVolume();
+    // calculate phi and dphi for this element
+    FEType fe_type(Utility::string_to_enum<Order>("first"),
+                   Utility::string_to_enum<FEFamily>("lagrange"));
+    std::unique_ptr<FEBase> fe(FEBase::build(1, fe_type));
+    fe->attach_quadrature_rule(const_cast<QBase *>(_assembly.qRule()));
+    const std::vector<RealGradient> * tangents = &fe->get_dxyzdxi();
+    unsigned side = 0;
+    fe->reinit(elem_ptr, side);
+    for (unsigned i = 0; i < tangents->size(); i++)
+      _slave_tangent += (*tangents)[i];
+  }
+
+  _slave_tangent /= _slave_tangent.norm();
+  _current_elem_volume /= elems.size();
```

Comment:
Lines 156 and 271 don't seem consistent.

Reply:
What would be a better way to get the element length/volume?